x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java

```diff
@@ -0,0 +1,252 @@
+        RoleReference.BwcApiKeyRoleReference bwcApiKeyRoleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        final List<RoleDescriptor> roleDescriptors = apiKeyService.parseRoleDescriptors(
+            bwcApiKeyRoleReference.getApiKeyId(),
+            bwcApiKeyRoleReference.getRoleDescriptorsMap()
+        );
+        final RolesRetrievalResult rolesRetrievalResult = new RolesRetrievalResult();
+        rolesRetrievalResult.addDescriptors(Set.copyOf(roleDescriptors));
+        listener.onResponse(rolesRetrievalResult);
+    }
+
+    @Override
+    public void resolveServiceAccountRoleReference(
+        RoleReference.ServiceAccountRoleReference roleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        serviceAccountService.getRoleDescriptorForPrincipal(roleReference.getPrincipal(), listener.map(roleDescriptor -> {
+            final RolesRetrievalResult rolesRetrievalResult = new RolesRetrievalResult();
+            rolesRetrievalResult.addDescriptors(Set.of(roleDescriptor));
+            return rolesRetrievalResult;
+        }));
+    }
+
+    private void resolveRoleNames(Set<String> roleNames, ActionListener<RolesRetrievalResult> listener) {
+        roleDescriptors(roleNames, ActionListener.wrap(rolesRetrievalResult -> {
+            logDeprecatedRoles(rolesRetrievalResult.getRoleDescriptors());
+            final boolean missingRoles = rolesRetrievalResult.getMissingRoles().isEmpty() == false;
+            if (missingRoles) {
+                logger.debug(() -> new ParameterizedMessage("Could not find roles with names {}", rolesRetrievalResult.getMissingRoles()));
+            }
+            final Set<RoleDescriptor> effectiveDescriptors;
+            Set<RoleDescriptor> roleDescriptors = rolesRetrievalResult.getRoleDescriptors();
+            if (roleDescriptors.stream().anyMatch(RoleDescriptor::isUsingDocumentOrFieldLevelSecurity)
+                && DOCUMENT_LEVEL_SECURITY_FEATURE.checkWithoutTracking(licenseState) == false) {
+                effectiveDescriptors = roleDescriptors.stream()
+                    .filter(not(RoleDescriptor::isUsingDocumentOrFieldLevelSecurity))
+                    .collect(Collectors.toSet());
+            } else {
+                effectiveDescriptors = roleDescriptors;
+            }
+            logger.trace(
+                () -> new ParameterizedMessage(
+                    "Exposing effective role descriptors [{}] for role names [{}]",
+                    effectiveDescriptors,
+                    roleNames
+                )
+            );
+            effectiveRoleDescriptorsConsumer.accept(Collections.unmodifiableCollection(effectiveDescriptors));
+            logger.trace(
+                () -> new ParameterizedMessage("Building role from descriptors [{}] for role names [{}]", effectiveDescriptors, roleNames)
+            );
+            // TODO: why not populate negativeLookupCache here with missing roles?
```

Comment:
The existing code populates the negativeLookupCache [after](https://github.com/elastic/elasticsearch/blob/85d12e5b5b45021a91078b2fce3030b36ccc8f0b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L386-L388) the `Role` gets successfully built.

I chose to retain this behaviour. But I wonder why it cannot be populated earlier here straight after roleDescriptor retrevial since we already know what roles are missing at this point?

Reply:
I don't think there's any reason for that. I think it just ended up there after introducing the RoleRetrievalResult

Reply:
I am going to leave it as is for now because:
1) It retains the current behaviour 
2) I'd like to move the effective role descriptor checking (licensing) from the Resolver back to CompositeRolesStore. My thinking is the the Resolver should just resolve the role descriptors and the caller will handle both the license check and both negative and positive caching. But this change has a decent size and potentially delay the process of this PR. So I'd like to handle it in a follow-up (I added a task in https://github.com/elastic/elasticsearch/issues/80117).